### PR TITLE
Feature pgm

### PR DESCRIPTION
**Pull request by planes**
This PR doesn't add PGM capabilities directly. This fixes many issues in SUAVE to enable that code to work with SUAVE.

### CODE REVIEW

**Comment by planes** [comment]
Obviously cyclic imports are not good. Unfortunately my machine has no problem with them. So it's hard for me to test for them.

**Comment by planes** [comment]
Ugh, now I need to add coverage.

File: trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py [hunk 1]
```diff
@@ -68,9 +70,10 @@ def vortex_lift(state,settings,geometry):
             # Apply to wing lift
             wing_lift[Mc < 1.0] = vortex_cl[Mc < 1.0]
         
-        wings_lift += wing_lift
+        # THIS CANNOT WORK
+        #wings_lift += wing_lift
```
Comment (mclarke2) [inline trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py:74]: Does this need to be made an issue and resolved? If so create one

File: trunk/SUAVE/Methods/Flight_Dynamics/Dynamic_Stability/compute_dynamic_flight_modes.py [hunk 2]
```diff
@@ -126,9 +126,9 @@ def compute_dynamic_flight_modes(results,aircraft,flight_conditions,cases):
     for wing in aircraft.wings:
         if wing.control_surfaces :
             for cs in wing.control_surfaces:
-                ctrl_surf =  wing.control_surfaces[cs]
+                ctrl_surf =  cs
```
Comment (mclarke2) [inline trunk/SUAVE/Methods/Flight_Dynamics/Dynamic_Stability/compute_dynamic_flight_modes.py:129]: this line can also now be removed by making the preceding line `for ctrl_surf in wing.control_surfaces`

File: regression/scripts/B737/mission_B737.py [hunk 3]
```diff
@@ -229,11 +221,10 @@ def simple_sizing(configs, analyses):
     weights = analyses.configs.base.weights
     breakdown = weights.evaluate()    
     
-    
     #compute centers of gravity
     #need to put here, otherwise, results won't be stored
-    compute_component_centers_of_gravity(base,compute_propulsor_origin=True)
-    compute_aircraft_center_of_gravity(base)
+    compute_component_centers_of_gravity(base)
+    base.CG()
```
[the remark this answers is not in the document]
Comment (planes) [reply regression/scripts/B737/mission_B737.py:227]: configs now calculate their own CGs. I could rename all CG functions to say something like: base.center_of_gravity()

File: regression/scripts/B737/mission_B737.py [hunk 4]
```diff
@@ -140,7 +132,7 @@ def base_analysis(vehicle):
 
     # ------------------------------------------------------------------
     #  Weights
-    weights = SUAVE.Analyses.Weights.Weights_Tube_Wing()
+    weights = SUAVE.Analyses.Weights.Weights_Transport()
```
[the remark this answers is not in the document]
Comment (planes) [reply regression/scripts/B737/mission_B737.py:135]: This is able to do more than tube and wing. I thought about it for a while and felt it was a significant departure from the prior methods that warranted the name

File: regression/scripts/noise_optimization/Procedure.py [hunk 5]
```diff
@@ -233,8 +233,8 @@ def estimate_clmax(nexus):
     # Condition to CLmax calculation: 90KTAS @ 10000ft, ISA
     conditions = Data()
     conditions.freestream = Data()
-    conditions.freestream.density           = 0.90477283
-    conditions.freestream.dynamic_viscosity = 1.69220918e-05
+    conditions.freestream.density           = [0.90477283]
```
[the remark this answers is not in the document]
Comment (planes) [reply regression/scripts/noise_optimization/Procedure.py:236]: because conditions in missions are usually vectors

File: trunk/SUAVE/Core/Data.py [hunk 6]
```diff
@@ -519,8 +521,17 @@ def deep_set(self,keys,val):
             for k in keys[:-1]:
                 data = data[k]
         
-        data[ keys[-1] ] = val
-        
+        if keys[-1][-1] ==']':
```
[the remark this answers is not in the document]
Comment (planes) [reply trunk/SUAVE/Core/Data.py:524]: Deep Set is mostly used in optimization cases where multiple values in a data structure require the same value simultaneously. Sometimes issues arise with the keys where you can end up with keys of keys that are put in brackets. I don't remember exactly where this problem arose, but I know it was a huge pain to figure out, and hence this change.

File: regression/scripts/B737/mission_B737.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (planes) [reply regression/scripts/B737/mission_B737.py]: I decided to make this more verbose. Pushed

File: regression/scripts/variable_cruise_distance/variable_cruise_distance.py [hunk 7]
```diff
@@ -38,14 +38,14 @@ def main():
...
-    distance_regression = 4178523.102867511
+    distance_regression = 4255970.731318368
     distance_calc       = results.conditions.frames.inertial.position_vector[-1,0]
     error_distance      = abs((distance_regression - distance_calc )/distance_regression)
     assert error_distance < 1e-6
     
     error_weight = abs(mission.target_landing_weight - results.conditions.weights.total_mass[-1,0])
     print('landing weight error' , error_weight)
-    assert error_weight < 1e-6
+    assert error_weight < 1e-3
```
[the remark this answers is not in the document]
Comment (planes) [reply regression/scripts/variable_cruise_distance/variable_cruise_distance.py]: Good catch, fixed.

File: trunk/SUAVE/Analyses/Stability/AVL.py [hunk 8]
```diff
@@ -424,7 +424,7 @@ def evaluate_conditions(self,run_conditions, trim_aircraft  ):
                 num_cs_on_wing = len(wing.control_surfaces)
                 num_cs +=  num_cs_on_wing
                 for cs in wing.control_surfaces:
-                    ctrl_surf = wing.control_surfaces[cs]     
+                    ctrl_surf = cs
```
[the remark this answers is not in the document]
Comment (planes) [reply trunk/SUAVE/Analyses/Stability/AVL.py]: Addressed

File: trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py (hunk 1 above)
Comment (planes) [reply trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py:74]: @timdmacdo, please provide your thoughts. Can this be made to work arbitrarily?

File: trunk/SUAVE/Methods/Flight_Dynamics/Dynamic_Stability/compute_dynamic_flight_modes.py (hunk 2 above)
Comment (planes) [reply trunk/SUAVE/Methods/Flight_Dynamics/Dynamic_Stability/compute_dynamic_flight_modes.py:129]: Addressed

File: trunk/SUAVE/Analyses/Weights/Weights_BWB.py [hunk 9]
```diff
@@ -58,62 +58,4 @@ def __defaults__(self):
...
-
-        Properties Used:
-        N/A
-            """           
-        self.mass_properties = self.vehicle.mass_properties
-        
-        return
-        
\ No newline at end of file
+        self.settings.empty = SUAVE.Methods.Weights.Correlations.BWB.empty
```
[the remark this answers is not in the document]
Comment (planes) [reply trunk/SUAVE/Analyses/Weights/Weights_BWB.py:61]: Addressed

File: regression/scripts/Vehicles/Beech_99.py [hunk 10]
```diff
@@ -38,13 +38,13 @@ def vehicle_setup():
     wing.aspect_ratio                        = wing.spans.projected**2/wing.areas.reference
     wing.symmetric                           = True
     wing.vertical                            = False
-    wing.origin                              = np.array([15.,0,0]) * Units.feet  
+    wing.origin                              = [[15.* Units.feet  ,0,0]]
```
[the remark this answers is not in the document]
Comment (planes) [reply regression/scripts/Vehicles/Beech_99.py:41]: Yes, to do truly arbitrary vehicles the data structures need to be consistent. This was one of the fixes necessary for the center_of_gravity function.

File: regression/scripts/Vehicles/Concorde.py [hunk 11]
```diff
@@ -254,7 +256,7 @@ def vehicle_setup():
     segment.root_chord_percent    = 14.5/14.5
     segment.dihedral_outboard     = 0.
     segment.sweeps.quarter_chord  = 63. * Units.deg
-    segment.thickness_to_chord    = 0.04
+    segment.thickness_to_chord    = 0.03
```
Comment (timdmacdo) [inline regression/scripts/Vehicles/Concorde.py:259]: Why were these thicknesses changed?

File: regression/scripts/aerodynamics/aerodynamics.py [hunk 12]
```diff
@@ -174,9 +174,9 @@ def main():
     #compute_aircraft_lift(conditions, configuration, geometry) 
     
     lift   = state.conditions.aerodynamics.lift_coefficient
-    lift_r =  np.array([-1.22715059, -0.48562864, -0.4312893 , -0.3442654 , -0.19773265,
-                         0.12732151,  0.37221063,  0.60102923,  0.90031948,  1.26978863,
-                         1.2846731 ])[:,None] 
+    lift_r =  np.array([-1.07644788, -0.42599003, -0.37832394, -0.30198719, -0.1734497 ,
```
Comment (timdmacdo) [inline regression/scripts/aerodynamics/aerodynamics.py:177]: Some of these changes are quite significant. What's the explanation here?

File: trunk/SUAVE/Analyses/Aerodynamics/Fidelity_Zero.py [hunk 13]
```diff
@@ -92,7 +92,6 @@ def __defaults__(self):
         compute.drag.parasite.propulsors.propulsor = Common.Drag.parasite_drag_propulsor
         compute.drag.parasite.pylons               = Common.Drag.parasite_drag_pylon
         compute.drag.parasite.total                = Common.Drag.parasite_total
-        compute.drag.induced                       = Common.Drag.induced_drag_aircraft
```
Comment (timdmacdo) [inline trunk/SUAVE/Analyses/Aerodynamics/Fidelity_Zero.py:95]: Why was this removed?

File: trunk/SUAVE/Components/Component.py [hunk 14]
```diff
@@ -45,6 +45,9 @@ def __defaults__(self):
             """         
         self.tag    = 'Component'
         self.origin = [[0.0,0.0,0.0]]
+        self.max_per_vehicle = 0
+        self.PGM_characteristics = []
```
Comment (timdmacdo) [inline trunk/SUAVE/Components/Component.py:49]: If the PGM files will be stored separately I think we may want to remove defaulted values from the main repo.

File: trunk/SUAVE/Components/Energy/Converters/Expansion_Nozzle.py [hunk 15]
```diff
@@ -155,6 +157,9 @@ def compute(self,conditions):
         Mach[i_high]  = 1.0*Mach[i_high]/Mach[i_high]
         P_out[i_high] = Pt_out[i_high]/(1.+(gamma[i_high]-1.)/2.*Mach[i_high]*Mach[i_high])**(gamma[i_high]/(gamma[i_high]-1.))
         
+        # A cap to make sure Mach doesn't go to zero:
+        Mach[Mach<=0.0] = 0.001
```
Comment (timdmacdo) [inline trunk/SUAVE/Components/Energy/Converters/Expansion_Nozzle.py:161]: I think it would be more appropriate to raise an error if this happens. Having assignments like this buried in the code can be confusing when trying to debug.

File: trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Lift/fuselage_correction.py [hunk 16]
```diff
@@ -41,7 +41,7 @@ def fuselage_correction(state,settings,geometry):
     wings_lift_comp = state.conditions.aerodynamics.lift_coefficient
     
     # total lift, accounting one fuselage
-    aircraft_lift_total = wings_lift_comp * fus_correction 
+    aircraft_lift_total = wings_lift_comp #* fus_correction
```
Comment (timdmacdo) [inline trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Lift/fuselage_correction.py:44]: It would be better to set this to 1 or just remove the functionality.

File: trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py (hunk 1 above)
Comment (timdmacdo) [reply trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py:74]: The version I currently use does this. What's the explanation for saying it can't work?

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Propulsion/compute_turbofan_geometry.py [hunk 17]
```diff
@@ -65,10 +59,13 @@ def compute_turbofan_geometry(turbofan, conditions):
 
     # pack
     turbofan.engine_length    = L_eng_m
-    turbofan.nacelle_diameter = nacelle_diameter
+    turbofan.nacelle_diameter = nacelle_diameter 
+    turbofan.inlet_diameter   = nacelle_diameter/np.sqrt(2.1)
```
Comment (timdmacdo) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Propulsion/compute_turbofan_geometry.py:63]: What is this 2.1 number from?

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Planform/rescale_non_dimensional.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+import SUAVE
+import numpy as np
+
+
+# ----------------------------------------------------------------------
+#  Set Origin Non-Dimensional
+# ----------------------------------------------------------------------
+
+def set_origin_non_dimensional(vehicle):
+    """ Places the origin of all components
```
Comment (timdmacdo) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Planform/rescale_non_dimensional.py:20]: Add more explanation here. It's not clear to me what this function is doing from the description.

File: trunk/SUAVE/Sizing/size_from_PGM.py [hunk 19]
```diff
@@ -0,0 +1,901 @@
+## @ingroup Sizing
```
Comment (timdmacdo) [inline trunk/SUAVE/Sizing/size_from_PGM.py:1]: Can this function be used without the PGM architecture?

File: regression/scripts/Vehicles/Concorde.py (hunk 11 above)
Comment (planes) [reply regression/scripts/Vehicles/Concorde.py]: Put back to normal, this accidentally got fed back in from an earlier version.

File: regression/scripts/aerodynamics/aerodynamics.py (hunk 12 above)
Comment (planes) [reply regression/scripts/aerodynamics/aerodynamics.py]: This should be corrected with reintroducing the fuselage correction.

File: trunk/SUAVE/Analyses/Aerodynamics/Fidelity_Zero.py (hunk 13 above)
Comment (planes) [reply trunk/SUAVE/Analyses/Aerodynamics/Fidelity_Zero.py:95]: The induced drag is computed by the VLM. This function is not of use for fidelity zero anymore.

File: trunk/SUAVE/Components/Component.py (hunk 14 above)
Comment (planes) [reply trunk/SUAVE/Components/Component.py:49]: Let's talk tomorrow.

File: trunk/SUAVE/Components/Energy/Converters/Expansion_Nozzle.py (hunk 15 above)
Comment (planes) [reply trunk/SUAVE/Components/Energy/Converters/Expansion_Nozzle.py]: I added a warning, there are similar warnings in other turbomachinery components.

File: trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Lift/fuselage_correction.py (hunk 16 above)
Comment (planes) [reply trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Lift/fuselage_correction.py]: I had been using the VLM including the fuselage before, but when I merged with develop that functionality was removed. So I put that back to normal. If we include the fuselage, the coefficient should go to 1.

File: trunk/SUAVE/Methods/Center_of_Gravity/compute_component_centers_of_gravity.py [hunk 20]
```diff
@@ -39,93 +44,137 @@ def compute_component_centers_of_gravity(vehicle, compute_propulsor_origin = Fal
...
+    hydraulics.mass_properties.center_of_gravity[0][0]         = 0.0       
+    
+    # Now the landing gear
+    
+    # Nose gear
+    nose_gear.origin[0][0]                                     = 0.25*nose_length
+    nose_gear.mass_properties.center_of_gravity[0][0]          = 0.0   
+    
+    # Main gear
+    moment_sans_main = vehicle.center_of_gravity()[0][0]*vehicle.sum_mass()-main_gear.mass_properties.mass
```
[the remark this answers is not in the document]
Comment (planes) [reply trunk/SUAVE/Methods/Center_of_Gravity/compute_component_centers_of_gravity.py]: Fixed

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Planform/rescale_non_dimensional.py (hunk 18 above)
Comment (planes) [reply trunk/SUAVE/Methods/Geometry/Two_Dimensional/Planform/rescale_non_dimensional.py]: Done

File: trunk/SUAVE/Sizing/size_from_PGM.py (hunk 19 above)
Comment (planes) [reply trunk/SUAVE/Sizing/size_from_PGM.py:1]: Yes, potentially. We can chat tomorrow about renaming it or adding it with tutorials or something.

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Propulsion/compute_turbofan_geometry.py (hunk 17 above)
Comment (planes) [reply trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Propulsion/compute_turbofan_geometry.py:63]: Honestly I don't remember. But, sqrt(2.1) is about 1.5. Which lines up with this source:
https://books.google.com/books?id=NeHoahlhCGMC&pg=PA185&lpg=PA185&dq=nacelle+diameter+to+inlet+diameter&source=bl&ots=DSj5NDl_E7&sig=ACfU3U3i6o9Ak6IOd5MqyvmieiLFKD7wbA&hl=en&sa=X&ved=2ahUKEwiu6My1suTpAhWPlHIEHQ7hB5QQ6AEwAHoECAgQAQ#v=onepage&q=nacelle%20diameter%20to%20inlet%20diameter&f=false

File: trunk/SUAVE/Analyses/Aerodynamics/Fidelity_Zero.py (hunk 13 above)
Comment (timdmacdo) [reply trunk/SUAVE/Analyses/Aerodynamics/Fidelity_Zero.py:95]: I don't we should remove it. Getting a reasonable value from the VLM requires extra geometry definition (i.e. twist) that may not be available.

File: trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py (hunk 1 above)
Comment (mclarke2) [reply trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Lift/vortex_lift.py]: @planes just bringing your attention to this

**Comment by planes** [comment]
@timdmacdo and @mclarke2, I think I've address all of your comments on this PR and travis is happy again.

**Comment by mclarke2** [comment]
Yes, Im actually reviewing right now

Matthew A. Clarke<http://www.matthewaclarke.com><https://profiles.stanford.edu/matthew-clarke>, E.I.T.
DARE Doctoral Fellow, Stanford University VPGE
Harold M. King Fellow, Tau Beta Pi Engineering Honor Society
Aerospace Design Laboratory<http://adl.stanford.edu/>
Ph.D. Candidate, Aeronautics and Astronautics | Stanford University
M.S. Aeronautics and Astronautics  '17 | Stanford University
B.S. Mechanical Engineering '16 | Howard University
tel: (202) 805-6599; email: mclarke2@stanford.edu<mailto:mclarke2@stanford.edu>



________________________________

File: trunk/SUAVE/Analyses/Aerodynamics/Supersonic_Zero.py [hunk 21]
```diff
@@ -62,11 +62,12 @@ def __defaults__(self):
         settings.trim_drag_correction_factor        = 1.02
         settings.wing_parasite_drag_form_factor     = 1.1
         settings.fuselage_parasite_drag_form_factor = 2.3
-        settings.aircraft_span_efficiency_factor    = 0.78
+        settings.aircraft_span_efficiency_factor    = 0.0
```
Comment (timdmacdo) [inline trunk/SUAVE/Analyses/Aerodynamics/Supersonic_Zero.py:65]: Should this be None?

Comment (planes) [reply trunk/SUAVE/Analyses/Aerodynamics/Supersonic_Zero.py:65]: No, it's a necessary parameter if you're using the supersonic zero analysis:
https://github.com/suavecode/SUAVE/blob/feature-PGM/trunk/SUAVE/Methods/Aerodynamics/Supersonic_Zero/Drag/induced_drag_aircraft.py#L53

Comment (planes) [reply trunk/SUAVE/Analyses/Aerodynamics/Supersonic_Zero.py:65]: I could have it default to 0.78, but that's rather arbitrary.

Comment (planes) [reply trunk/SUAVE/Analyses/Aerodynamics/Supersonic_Zero.py:65]: Wait, I take this all back. This parameter should be removed entirely.
